src/charm.py

```diff
@@ -1,104 +1,400 @@
+    # ====================
+
+    def _on_reload_pgbouncer_action(self, event) -> None:
+        """An action to explicitly reload pgbouncer.
+
+        TODO implement function stub
+        """
+        self._reload_pgbouncer()
+        event.set_results({"result": "pgbouncer has restarted"})
+
+    def _on_change_password_action(self, event) -> None:
+        """An action to update the password for a specific user.
+
+        TODO implement password encryption at this stage, before anything is stored.
+        TODO verify passwords are valid
+        """
+        username = event.params["username"]
+        users = self._get_userlist_from_container()
+        if username not in users:
+            event.set_results(
+                {
+                    "result": f"user {username} does not exist - use the get-users action to list existing users."
+                }
+            )
+            return
+        hash = hashlib.md5(event.params["password"].encode())
+        users[username] = hash.hexdigest()
+        self._push_container_config(users=users)
+        event.set_results({"result": f"password updated for user {username}"})
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
+    def _on_add_user_action(self, event) -> None:
+        """Event handler for add-user action.
 
-        Learn more about config at https://juju.is/docs/sdk/config
+        TODO implement password encryption at this stage, before anything is ever stored.
+        TODO verify passwords are valid
         """
-        current = self.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
+        users = self._get_userlist_from_container()
+        username = event.params["username"]
+        if username in users:
+            event.set_results({"result": f"user {username} already exists"})
+            return
 
-    def _on_fortune_action(self, event):
-        """Just an example to show how to receive actions.
+        hash = hashlib.md5(event.params["password"].encode())
+        users[username] = hash.hexdigest()
+        self._push_container_config(users=users)
+        event.set_results({"result": f"new user {username} added"})
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle actions, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the actions.py file.
+    def _on_remove_user_action(self, event) -> None:
+        """Event handler for remove-user action.
 
-        Learn more about actions at https://juju.is/docs/sdk/actions
+        TODO This only removes the user from userlist.txt, and not from the juju config. Do we need
```

Comment:
This has been written into a Jira spike ticket, but I think it's important to leave this particular condition in the code, since it also documents unexpected behaviour.

Reply:
Sounds good to me, I can get behind that. To make things more clear for others perhaps you can keep the parts that document unexpected behaviour and remove the parts that are notes for yourself. 

Maybe something like:

TODO This should use Juju Secrets once available. As of now this only removes the user from userlist.txt, and not from the juju config. Therefore in the current implementation the password is updated, so the user is inaccessible and thus "removed".